[BG fields] Add refactored TWTS laser

BeyondEspresso opened the pull request:
By re-implementing the TWTS laser as functors, this request concludes a series based on the original pull request #597. Refactoring its separate features led to the subsequent pull requests #600 (BG Fields in mySimulation.hpp, merged), the #664 (Complex class in libPMacc, open) and now this one.

* Compared to #597, the implementation also includes the correct back-rotation of vector components on the staggered components of the field grids, improved support for 2D and .
* Functionality tests with and without fields show the same results than a parallel implementation without functors, which directly proceeded #597.
* Performance tests show a lower performance (~10% without particles and currents, ~1% with particles and currents) than the reference implementation. This performance *probably* arises from an increased number of float_64 variables. This penalty in performance is currently acceptable.
* The full sources and documentation of the equations and program that generated the core code for the fields are now at Klaus for cross-checking. However, since this chross-check is done independently and will take some time (probably ~1 month), the focus of this pull request will be more on the implementation aspects.
* The code includes the current unmerged status of #664.
* **Future:** This code will be updated as soon as the *const expression* feature becomes available in CUDA and PIConGPU. The aim of this update will be improved speed and a further generalization of the field rotation routines as functors for general use in PIConGPU/libPMacc.

Happy code ripping! :yum:

### To Do

- [x] rebase against `dev` as soon as #664 is merged

BeyondEspresso commented:
The 2D case is currently not passing.

ax3l commented:
thank you for the pull request, we will do a detailed review next week.

Have a good weekend! :)

BeyondEspresso commented:
Above fix is very dirty :poop: . Suggestions on improvements, without resorting to pre-compiler directives, are welcome :innocent: .

bussmann commented:
Nice!

ax3l commented:
it's still a bit hard to review due to the missing rebase against the merged complex pull and the huge number of single commits.

I will try a first swipe, but no guarantee :)

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,659 @@
+        }
+        
+        template<>
+        HDINLINE float3_X
```
since you return the result in `SI` pls use `float3_64` here, too.

BeyondEspresso replied:
Since the result is a normalized quantity the datatype is correct, but I renamed the function.

ax3l replied:
makes sense, well done.

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,659 @@
+        }
+        
+        template<>
+        HDINLINE float3_X
```
`float3_64`

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,659 @@
+            return float3_X( float_X(0.), float_X(0.), precisionCast<float_X>( calcTWTSEx(eFieldPositions_SI[2],time)/unitField[2] ) );
+        }
+        
+        HDINLINE float3_X
```
`float3_64`

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,659 @@
+         * \param pos Spatial position of the target field.
+         * \param time Absolute time (SI, including all offsets and transformations) for calculating the field */
+        HDINLINE float_64
+        TWTSFieldE::calcTWTSEx( const float3_64& pos, const float_64 time) const
```
a general remark: if you calculate the factors below here all accordingly, your calculation can be speeded up by a factor 3 since you can in principle use `float_X` for your operations.

ax3l commented on src/picongpu/include/simulationControl/MySimulation.hpp:
```diff
@@ -342,10 +342,10 @@ class MySimulation : public SimulationHelper<simDim>
-                            step, fieldBackgroundE::InfluenceParticlePusher);
-            (*pushBGField)( fieldB, nvfct::Sub(), fieldBackgroundB(fieldB->getUnit()),
-                            step, fieldBackgroundB::InfluenceParticlePusher);
+            (*pushBGField)( fieldE, nvfct::Sub(), FieldBackgroundE(fieldE->getUnit()),
```
that change is correct, but since it is highly risky for merge conflicts and is unrelated to the functionality of the pull please do not change it at the same time as you add twts.

BeyondEspresso replied:
As discussed outside this review, this change is not so critical that we need to open an independent issue.

ax3l commented on src/picongpu/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -33,7 +37,23 @@ namespace picongpu
         const float3_64 unitField;
-        HDINLINE fieldBackgroundE( const float3_64 unitField ) : unitField(unitField)
+
+        /* TWTS E-fields need to be initialized on host, so they can look up global grid dimensions. */
```
that is a nice example.

the only thing: it should not be part of the `defaul` files in `src/picongpu/include/simulation_defines/`.
please just move it to either `examples/Bunch` or `examples/LaserWakefield` and we can add a specific test there that activates and uses it.

BeyondEspresso commented:
Pull request repository is rebased.

BeyondEspresso commented:
In addition to your suggestions I further cleaned up code and interface. Now the code precision can be changed on an individual basis and the coordinate normalization should make it possible now to use the speedier float_X datatype.

Since this round of refactoring has now ended, please go ahead with the review and hack this to pieces! :smile:

ax3l commented:
sounds great, we will have a look! :sparkles:

ax3l commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+/** Load pre-defined templates */
+#include "fields/background/templates/bgrTWTS.def"
+
+#ifndef PARAM_INCLUDE_FIELDBACKGROUND
```
@BeyondEspresso you have become a very gentle developer in a steep curve, I like that! Nice and clean. :+1:

ax3l commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+                /* propagation speed of overlap [speed of light]. */
+                1.0,
+                /* manual time delay [s] if auto_tdelay is false */
+                ( 39.3e-6 )/(::picongpu::SI::SPEED_OF_LIGHT_SI),
```
I am a bit suprised the `::picongpu::` prefix is necessary - it should be possible to skip that in this scope.

ax3l commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+            const double WAVE_LENGTH_SI = 0.8e-6;
+            
+            /** UNITCONV */
+            /* const double UNITCONV_Intens_to_A0 = ::picongpu::SI::ELECTRON_CHARGE_SI
```
same here, we should be able to skip `::picongpu::`

ax3l commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+                             1.0/unitField[2]);
+            
+            const float3_X AMPLITUDE_SI = precisionCast<float_X>(
+                                float_64(_A0 * UNITCONV_A0_to_Amplitude_SI) * invUnitField );
```
is the point-wise operation `/ unitField` not defined?

ax3l replied:
this should do it (the second cast is not necessary):
```c++
const float3_X AMPLITUDE_SI = precisionCast<float_X>(
      _A0 * UNITCONV_A0_to_Amplitude_SI / unitField );
```

BeyondEspresso replied:
That's what I tried first. Maybe the point-wise `/unitField` does not go well with `double` variables.

ax3l replied:
@psychocoderHPC interesting. is that a problem specific to 64bit?

ax3l replied:
side note: something is wrong. either the name `AMPLITUDE_SI` or the `float3_X`.

psychocoderHPC replied:
I looked to the code and see that not all vector/native type operations are implemented.
I opend the issue  #738

@BeyondEspresso: Please open a issue if you found such things were you need a workaround to solve a problem. After the issue is opened implement your work around and mark the codepart with `/*\todo*/`

BeyondEspresso replied:
Name was wrong.

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.def:
```diff
@@ -0,0 +1,263 @@
+            the laser gradually enters the simulation volume? [Default: TRUE] */
+            const PMACC_ALIGN(auto_tdelay,bool);
+            /* Number of field components used in the simulation. [Default: 3 for both 2D and 3D] */
+            const static int numComponents=DIM3;
```
We could better use this from `FieldE` via:
```c++
#include "fields/Fields.def"
// ...
const static int numComponents = FieldE::numComponents;
```

The problem right now is, that the numComponents attribute is right now added in the field's `.hpp` file. @psychocoderHPC should we move that general parts to the field's `.def` file?

psychocoderHPC replied:
IMO: you can include `Fields.hpp` without side effects. `Fields.def` is only needed for the forward declaration inside `FieldE.hpp`,`FieldB.hpp`

ax3l replied:
I am not sure about that: the `.hpp` files include stuff like `#include "simulation_defines.hpp"` but this very file right now is a `def` file which is used in `param` files (besides that a `Fields.hpp` does not exist due to that reason).

ax3l replied:
nevertheless, I think that requires more restructuring so for me keep it fixed to `DIM3` even if `DIM` is actually a `position` notation and not a component notation (and causes a lot of confusion to readers and non-physicists).

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.def:
```diff
@@ -0,0 +1,263 @@
+            template<unsigned T_dim>
+            HDINLINE float_64 getTime_SI(const uint32_t currentStep) const;
+
+            HDINLINE PMacc::math::Vector<float3_X,numComponents>
```
Even if I think we should be able to avoid that `dummyCast` - here is the right place to add it's doxygen strings.

psychocoderHPC replied:
all `dummyCast`s are defined and implemented in `TWTSFieldE` and `TWTSFieldB`. As I can see the implementations were equal. Thus it would be nice to move the function out of the classes and only implement them once.

BeyondEspresso replied:
The implementations differ at `FieldE::numComponents`. In principle we could also move this component out of the class. However this touches upon the many parallels between the TWTS E-field and B-field class. It could be interesting to make have a base class from which aforementioned classes inherit. This would limit code doubling. What do you think?

ax3l replied:
sounds good :+1: no need to inherit, just add a independent auxiliary class/functors in a `detail` namespace

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+            {
+                PMacc::math::Vector<float3_X, FieldE::numComponents> result;
+
+                for( uint32_t i = 0; i < numComponents; ++i ) // cellIdx Ex, Ey and Ez
```
signed-unsigned compare, pls use `int`, too

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+            /* Note: These objects cannot be instantiated on CUDA GPU device. Since this is done
+                     on host (see fieldBackground.param), this is no problem. */
+            const SubGrid<simDim>& subGrid = Environment<simDim>::get().SubGrid();
+            halfSimSize=subGrid.getGlobalDomain().size / 2;
```
spaces around ` = ` :)

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+                
+                PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI;
+                
+                for( uint32_t i = 0; i < FieldE::numComponents; ++i ) // cellIdx Ex, Ey and Ez
```
signed-unsigned compare, too - see above

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+                     *  use the angle phi to determine the required amount of pulse front tilt.
+                     *  RotationMatrix[PI/2+phi].(y,z) (180Deg-flip at phi=90Deg since coordinate
+                     *  system in paper is oriented the other way round.) */
+                    eFieldPositions_SI[i] = float3_64( (eFieldPositions_SI[i]).x(),
```
`()` in `(eFieldPositions_SI[i])` can probably be removed

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+                
+                for( uint32_t i = 0; i < FieldE::numComponents; ++i ) // cellIdx Ex, Ey and Ez
+                {
+                    eFieldPositions[i]   += ( dummyCast( floatD_X(cellIdx) ) - laserOrigin );
```
the only reason the `dummyCast` is needed here is, because `laserOrigin` is not correctly defined as `floatD_X`.

BeyondEspresso replied:
No, the reason for `dummyCast` is that I am always constructing a 3D position vector (eFieldPositions[i]). The problem is that the 3D and 2D code sections have to parse for the respective other case (2D and 3D).

ax3l replied:
I know and it's not wrong but sub-ideal.
positions should always be of type `floatD_` but I know that continues in other parts of the implementation.

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+                                                 * unit_length;
+                
+                /* TWTS laser coordinate origin is centered transversally and defined longitudinally
+                 * by the laser center (usually maximum of intensity) in y. */
```
add: `in cells`

BeyondEspresso replied:
Can you explain what you mean by this comment?

ax3l replied:
basically, I just want that you add that you implicitly counted the "origin" in cells since it is not mentioned or reflected by the variable name.

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+            /* For the Yee-Cell shifted fields, obtain the fractional cell index components and add
+             * that to the total cell indices. The physical field coordinate origin is transversally
+             * centered with respect to the global simulation volume. */
+            PMacc::math::Vector<float3_X, FieldE::numComponents> eFieldPositions =
```
`eFieldPositions` should be, e.g. as in `include/fields/numericalCellTypes/YeeCell.hpp`, of type `PMacc::math::Vector<floatD_X,DIM3>`.

That avoids the `dummyCast`.

BeyondEspresso replied:
This would not work, because then eFieldPositions would be a vector of 2D vectors in the 2D case. It should be 3D, since I am always calling the calcTWTSEx(), which requires a 3D position vector. Since I have to rotate the coordinate system and set one coordinate to zero (except potentially for a intraCell-shift), I have to differentiate between the coordinates beyond shuffling components. Furthermore without pre-compiler directives all code needs to parse for all dimensions. Here I avoid this dilemma by providing a way to cast 2D vectors to 3D vectors via `dummyCast`. Better solutions are welcome.

ax3l replied:
I know that you *require* the 3D position vector, I am just trying to help you solving the parts of the code that *force positions* to be 3D. Which is indeed only necessary due to your implementation.

Of course, that is not to confused with the 3-components of the field itself, which is independent of the dimensionality of the simulation always 3 (see `PMacc::math::Vector<floatD_X,DIM3>`).

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+            
+            PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI;
+            
+            for( uint32_t i = 0; i < FieldE::numComponents; ++i ) /* cellIdx Ex, Ey and Ez */
```
signed to unsigned comparison

ax3l commented:
general remark: can you please use spaces consequently around assignment operations such as `a = b / c;`? :)

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+                 *  By --> By
+                 *  Bz --> -Bx
+                 */
+                eFieldPositions_SI[i] = float3_64( -(eFieldPositions_SI[i]).z(),
```
I think we have a `TwistVector` operation for shuffles like that. @psychocoderHPC might now that.

BeyondEspresso replied:
`TwistVector` shuffles, but does not include signs. I discussed this with @Heikman . Furthermore I would argue that `TwistVector` does not greatly improve readability in this specific case. Hence this would be just an other dependency with no gain besides syntactic sugar.

ax3l replied:
well but that is easy to solve: shuffle and point-wise multiply with `float3_64( -1., 1., 1.)`. With that you can avoid the loop and the forcing of `float3` for positions.

ax3l commented on src/picongpu/include/fields/background/templates/bgrTWTS.hpp:
```diff
@@ -0,0 +1,953 @@
+         * \param time Absolute time (SI, including all offsets and transformations) for calculating
+         *             the field */
+        HDINLINE TWTSFieldE::float_T
+        TWTSFieldE::calcTWTSEx( const float3_64& pos, const float_64 time) const
```
any `pos` should always be of type `floatD_`

BeyondEspresso replied:
This one is different, because it always takes a 3D vector. The dimensional specialization is done in `getTWTSEfield_Normalized<simDim>()`. This keeps the core functions clean from clutter.

ax3l replied:
I see the way it is implemented and I am trying to reduce the "I force everywhere `float3` clutter".

To reduce it to the implementation of this function (and to avoid writing it in a less explicit way) one could also handle the `floatD_64` to `float3_64` dummy cast *just inside this function*.

ax3l commented:
I think the pull is already in a nice shape. The question is how consequently we want-to/can clean up internals such as `calcTWTSEx` to use `floatD_64` for positions and to write the following formulas accordingly to not differenciate between `x` and `z` as done for now.

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+                             1.0/unitField[1],
+                             1.0/unitField[2]);
+            
+            const float3_X AMPLITUDE_SI = precisionCast<float_X>(
```
please use the name `amplitude_SI` or `amplitude_si` instead of `AMPLITUDE_SI`
A typical C/C++ code convention is to write only defines with upper case letters.

In PIConGPU we sometimes use *upper case letters* for global param/unitless variables inside namespace.

ax3l replied:
yes, he took it from the `laserConfig.param` files to be consistent.

ax3l replied:
offline discussion -> post-poned to #739

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+            /* const double _A0  = _PEAK_INTENSITY_SI * UNITCONV_Intens_to_A0; */
+
+            /** unit: none */
+            const double _A0  = 1.0;
```
please do not begin variables with `_`. The `_` is reserved in `C++` for identifier.

ax3l replied:
also consistent with `laserConfig.param` -> I would not change it now but change it consistently in both later on.

ax3l replied:
offline discussion -> post-poned to #739

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,222 @@
+        const float3_64 unitField;
+        HINLINE FieldBackgroundB( const float3_64 unitField ) : unitField(unitField),
+        
+            twtsFieldB( 30.0e-6,  /* focus_y [m], the distance to the laser focus in y-direction */
```
please move your comment above the number to be consistent with the following lines.

BeyondEspresso commented:
Thank you all for your constructive criticism. Here is the next iteration with your suggestions included.

I got rid of all dummyCast()-functions, so that 2D/3D decisions by the compiler are separated by function and variable interfaces and thus appear to be more *natural*.

Furthermore I got rid of duplicate code and reduced the size of the individual code chunks by distributing the code to separate files. Nevertheless, the number of lines has increased due to the increased number of copyright disclaimers.

The code should be in decent shape by now, but feel free to point out any wrongdoings on my part.  :volcano:

ax3l commented:
needs rebase against dev due to merge conflict

ax3l commented:
feel free to squash some of the `28` commits before rebasing on `dev` to make your live easier :)

BeyondEspresso commented:
Rebased against `dev`. Went surprisingly smooth. :yum:

There was a small hiccup in
`src/picongpu/include/simulation_defines/_defaultUnitless.loader` in 0c25239
where Git wanted to know that it should indeed use all `#include`-statements.
In addition I squashed all intermediate bugfixing and refactoring commits.

ax3l commented:
Fantastic, thank you! I'll review it asap.

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,225 @@
+        static const bool InfluenceParticlePusher = PARAM_INCLUDE_FIELDBACKGROUND;
+
+        /* We use this to calculate your SI input back to our unit system */
+        const float3_64 unitField;
```
please use `PMACC_ALIGN` to avoid side effects.

`PMACC_ALIGN(unitField,const float3_64);`

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,225 @@
+
+        /* TWTS E-fields need to be initialized on host,
+         *  so they can look up global grid dimensions. */
+        const templates::twts::EField twtsFieldE;
```
please use `PMACC_ALIGN`

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,225 @@
+
+        /* TWTS B-fields need to be initialized on host,
+         * so they can look up global grid dimensions. */
+        templates::twts::BField twtsFieldB;
```
please use `PMACC_ALIGN`

psychocoderHPC commented on examples/Bunch/include/simulation_defines/param/fieldBackground.param:
```diff
@@ -0,0 +1,225 @@
+        static const bool activated = false;
+
+        /* We use this to calculate your SI input back to our unit system */
+        const float3_64 unitField;
```
please use `PMACC_ALIGN`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+        /* Since we rotated all position vectors before calling calcTWTSBy and calcTWTSBz,
+         * we need to back-rotate the resulting B-field vector. */
+        /* RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the Field-Vektors. */
+        const float_64 By_rot = -sin(+phi)*By_By+cos(+phi)*Bz_By;
```
please use `pmMath::sin` and `pmMath::cos`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+         * we need to back-rotate the resulting B-field vector. */
+        /* RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the Field-Vektors. */
+        const float_64 By_rot = -sin(+phi)*By_By+cos(+phi)*Bz_By;
+        const float_64 Bz_rot = -cos(+phi)*By_Bz-sin(+phi)*Bz_Bz;
```
please use `pmMath::sin` and `pmMath::cos`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+         * need to back-rotate the resulting B-field vector. Now the rotation is done
+         * analogously in the (y,x)-plane. (Reverse of the position vector transformation.) */
+        /* RotationMatrix[-(PI/2+phi)].(By,Bx) */
+        const float_64 By_rot = -sin(phi)*By_By+cos(phi)*Bx_By;
```
please use `pmMath::sin` and `pmMath::cos`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+        /* RotationMatrix[-(PI/2+phi)].(By,Bx) */
+        const float_64 By_rot = -sin(phi)*By_By+cos(phi)*Bx_By;
+        /* for rotating back the Field-Vektors.*/
+        const float_64 Bx_rot = -cos(phi)*By_Bx-sin(phi)*Bx_Bx;
```
please use `pmMath::sin` and `pmMath::cos`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+        /* propagation speed of overlap normalized to the speed of light [Default: beta0=1.0] */
+        const float_T beta0 = float_T(beta_0);
+        const float_T phiReal = float_T(phi);
+        const float_T alphaTilt = atan2(float_T(1.0)-beta0*cos(phiReal),beta0*sin(phiReal));
```
please use `pmMath::*`

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,404 @@
+        const float_T t = float_T(time/UNIT_TIME);
+                        
+        /* Shortcuts for speeding up the field calculation. */
+        const float_T sinPhi = sin(phiT);
```
please use `pmMath::*` for all math functions bellow

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/EField.tpp:
```diff
@@ -0,0 +1,242 @@
+        /* propagation speed of overlap normalized to the speed of light [Default: beta0=1.0] */
+        const float_T beta0 = float_T(beta_0);
+        const float_T phiReal = float_T(phi);
+        const float_T alphaTilt = atan2(float_T(1.0)-beta0*cos(phiReal),beta0*sin(phiReal));
```
please use `pmMath::*` for all math functions in this file

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/GetInitialTimeDelay_SI.tpp:
```diff
@@ -0,0 +1,159 @@
+             * projection we calculate the y-distance walkoff of the TWTS-pulse.
+             * The abs()-function is for correct offset for -phi<-90Deg and +phi>+90Deg. */
+            const float_64 y1 = float_64(halfSimSize[0]
+                                *picongpu::SI::CELL_WIDTH_SI)*abs(cos(eta));
```
please use `pmMath::*` for all math functions in this file

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/RotateField.tpp:
```diff
@@ -0,0 +1,127 @@
+                fieldPosVector.x(),
+               -pmMath::sin(AngleType(phi))*fieldPosVector.y()
+                    -pmMath::cos(AngleType(phi))*fieldPosVector.z() ,
+               +pmMath::cos(AngleType(phi))*fieldPosVector.y()
```
Is the plus `+` needed?

BeyondEspresso replied:
No, but explicit signs increase the readability of the math here.

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/RotateField.tpp:
```diff
@@ -0,0 +1,127 @@
+             return result(
+                -pmMath::sin(AngleType(phi))*fieldPosVector.y()
+                    -pmMath::cos(AngleType(phi))*fieldPosVector.x() ,
+                +pmMath::cos(AngleType(phi))*fieldPosVector.y()
```
Is the plus `+` needed?

BeyondEspresso replied:
Readability. Also it shows that among the 3 `-` signs it is not a typo, but intended.

psychocoderHPC commented on src/picongpu/include/fields/background/templates/TWTS/RotateField.tpp:
```diff
@@ -0,0 +1,127 @@
+             *  y -->  y
+             *  z --> -x (Since z=0 for 2D, we use the existing
+             *            TWTS-field-function and set -x=0)
+             *  Ex --> Ez (--> Same function values can be used in 2D,
```
I am confused that the description contains `E-Field` and `B-Field` but this class nothing knows about field types (only about a field position).
@ax3l please check if it is more clear for you than for me

BeyondEspresso replied:
The three comment lines on the field components are not strictly needed here. However, since the field components transform in the same way than the position vectors it is IMHO good to make this explicit in at least one comment section and not keep everything separated.
The intention is to tell everything there is to know about the rotational transforms of the TWTS laser in one place. After splitting up the code in separate classes this might look funny, but there should be at least one place close to the implementation where the details are *all* mentioned.
Maybe I should expand the comment in the code to make this more clear?

ax3l replied:
yes maybe, the comment is also not very clear to me and we should not assume an always staggered "Yee" in such an implementation.

BeyondEspresso replied:
Can you elaborate on what confuses you in the comment? Additionally, at no point in this implementation I assume a Yee-Cell grid. I only used this phrase in the comment, because it sounds more familiar than "optional intra-cell offset due to staggered grid".

ax3l replied:
I do not understand what `Ex --> Ez (--> Same function values can be used in 2D but with Yee-Cell-Positions for Ez.)` shall tell me and why it is specific to `Yee` since PIConGPU supports arbitrary positions for fields in cells, not only Yee staggering.

ax3l replied:
pls address *this* question :)

BeyondEspresso replied:
As I told you, it is *not* specific to the Yee staggering. I am going to use phrases such as intra-cell offset and staggered grids instead.

BeyondEspresso commented:
@psychocoderHPC 's suggestions are now online.

ax3l commented:
small reminder: [spaces around operators](https://github.com/ComputationalRadiationPhysics/picongpu/pull/704#issuecomment-76929878) increase readability (and are part of our coding guidelines).

BeyondEspresso commented:
Fixed this.
Wrote  #766 as constructive criticism on the "spacings around division operator coding guideline".

ax3l commented:
@BeyondEspresso great work!

@psychocoderHPC since this is a bigger pull request and I think it is ready for a merge for now - are you d'accord? :)

ax3l commented on src/picongpu/include/fields/background/templates/TWTS/BField.tpp:
```diff
@@ -0,0 +1,409 @@
+        const float_64 Bz_Bz = calcTWTSBz(pos[2], time);
+        /* Since we rotated all position vectors before calling calcTWTSBy and calcTWTSBz,
+         * we need to back-rotate the resulting B-field vector. */
+        /* RotationMatrix[-(PI/2+phi)].(By,Bz) for rotating back the Field-Vektors. */
```
Vectors ;)

BeyondEspresso replied:
@ax3l Note that this is currently unique code that applies to the B-field only (because of intra-cell position offsets). My intention is to generalize this in the upcoming feature update when I introduce different polarizations for the field.

ax3l replied:
:)

BeyondEspresso commented:
Ok, I found and fixed another bug in the 2D section. In addition I did some more polishing on the comments. Let's merge this now. :wink:

BeyondEspresso commented:
Bad bugs kick the bucket last...

ax3l commented:
merge-able now? ;)
